### PR TITLE
Added Facebook picture filters

### PR DESCRIPTION
**Pull request by rpg600**

### CODE REVIEW

File: src/AppBundle/Facebook/ImageFilter.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        $picture = $this->imagine->open($picture);
+
+        $generatedPictures = [];
+
+        foreach ($this->watermarks as $watermark) {
+            $generatedPictures[] = base64_encode($this->addWatermark($watermark, $picture)->get('jpeg'));
+        }
+
+        return $generatedPictures;
```
Comment (HeahDude) [inline src/AppBundle/Facebook/ImageFilter.php]: A micro optim of PHP7 is when the array is only initialized in the loop, so you can remove the one line 31 and do instead:
```php
return $generatedPictures ?? [];
```
The compiled code is more performant and may prevent a memory access.

File: app/Resources/views/facebook/index.html.twig [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{% extends 'base.html.twig' %}
+
+{% block content %}
+    <main class="facebook">
+        {% include 'facebook/header.html.twig' %}
```
Comment (hhamon) [inline app/Resources/views/facebook/index.html.twig]: Use the `include()` function instead please to be consistent with all templates.

File: app/Resources/views/facebook/show.html.twig [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{% extends 'base.html.twig' %}
+
+{% block content %}
+    <main class="facebook">
+        {% include 'facebook/header.html.twig' %}
```
Comment (hhamon) [inline app/Resources/views/facebook/show.html.twig]: Same here.

File: src/AppBundle/Controller/FacebookController.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * @Route("/auth", name="app_facebook_auth")
+     * @Method("GET")
+     */
+    public function authAction(): RedirectResponse
+    {
+        $fb = $this->get('app.facebook.api');
+        $helper = $fb->getRedirectLoginHelper();
+
+        $redirectUrl = $this->generateUrl(
```
Comment (hhamon) [inline src/AppBundle/Controller/FacebookController.php]: Can you move this directly in the last line?

File: src/AppBundle/Facebook/ImageFilter.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+use Imagine\Image\ImageInterface;
+use Imagine\Image\ImagineInterface;
+use Imagine\Image\Point;
+use Symfony\Component\Finder\Finder;
+
+class ImageFilter
+{
+    private $imagine;
+
```
Comment (hhamon) [inline src/AppBundle/Facebook/ImageFilter.php]: remove extra blank like.

File: src/AppBundle/Facebook/ImageFilter.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    private $watermarks;
+
+    public function __construct(ImagineInterface $imagine, string $watermarksPath)
+    {
+        $this->imagine = $imagine;
+
```
Comment (hhamon) [inline src/AppBundle/Facebook/ImageFilter.php]: remove extra blank like.

File: src/AppBundle/Facebook/ImageFilter.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Applies watermarks on the given picture path
+     * and returns an array of base64 encoded jpeg pictures.
+     */
+    public function applyWatermarks(string $picture): array
+    {
+        $picture = $this->imagine->open($picture);
+
+        $generatedPictures = [];
+
```
Comment (hhamon) [inline src/AppBundle/Facebook/ImageFilter.php]: remove extra blank like.

File: src/AppBundle/Facebook/ImageFilter.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+{
+    private $imagine;
+
+    private $watermarks;
+
+    public function __construct(ImagineInterface $imagine, string $watermarksPath)
+    {
+        $this->imagine = $imagine;
+
+        $this->watermarks = $this->generateWatermarks($watermarksPath);
```
Comment (hhamon) [inline src/AppBundle/Facebook/ImageFilter.php]: I wouldn't do this in the construct method but when needed only. The construct method must never perform any complex logic such as hitting the filesystem. Construct method should only initialize the state of the object.

File: src/AppBundle/Facebook/ImageFilter.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+        $finder = new Finder();
+        $finder->files()->in($path);
+
+        if (0 === count($finder)) {
+            throw new \RuntimeException(sprintf('No watermark file found in directory %s', $path));
+        }
+
+        $watermarks = [];
+
```
Comment (hhamon) [inline src/AppBundle/Facebook/ImageFilter.php]: remove extra blank like.

File: src/AppBundle/Facebook/ImageFilter.php [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            $generatedPictures[] = base64_encode($this->addWatermark($watermark, $picture)->get('jpeg'));
+        }
+
+        return $generatedPictures;
+    }
+
+    private function addWatermark(ImageInterface $watermark, ImageInterface $picture): ImageInterface
+    {
+        $watermark->resize($picture->getSize());
+
```
Comment (hhamon) [inline src/AppBundle/Facebook/ImageFilter.php]: remove extra blank like.

File: src/AppBundle/Facebook/ImageFilter.php (hunk 1 above)
Comment (rpg600) [inline src/AppBundle/Facebook/ImageFilter.php]: Nice 👍

File: src/AppBundle/Facebook/ImageFilter.php [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * Generate an array of watermarks image for the given path.
+     */
+    private function generateWatermarks(): array
+    {
+        if (!is_dir($this->path)) {
+            throw new \RuntimeException(sprintf('The %s directory does not exist.', $this->path));
+        }
+
+        $finder = new Finder();
```
Comment (HeahDude) [inline src/AppBundle/Facebook/ImageFilter.php]: Why don't you inject it?

**Pull request merged**